lib/Sema/CSSimplify.cpp

```diff
@@ -1125,6 +1125,53 @@ ConstraintSystem::TypeMatchResult constraints::matchCallArguments(
       // they mirror parameter flags to ease diagnosis.
       assert(!argsWithLabels[argIdx].isAutoClosure() ||
              isSynthesizedArgument(argument));
+      
+      // For argument conversion key path, when we have an unresolved root type
+      // key path e.g. \.value with a PartialKeyPath<Root> argument, let's not
+      // attempt to bind the key path type variable type to it and instead, iff
+      // we can, try to find a more specific type for this type variabl based
+      // on the member type.
+      if (auto *typeVar = argTy->getAs<TypeVariableType>()) {
```

Comment:
Okay, as promised - looks like this is not actually a problem with bindings or matching, it's rooted in `simplifyKeyPathConstraint`.

If you look at `let _: PartialKeyPath<SR12390> = \. value` - it works just fine today because there is this bit of code which makes it possible - https://github.com/apple/swift/blob/master/lib/Sema/CSSimplify.cpp#L7534. `tryMatchRootAndValueFromType` has `allowPartial` flag which is only `true` when type is contextual, for everything else it's false - that's why when parameter type is a partial key path `simplifyKeyPathConstraint` fails.

The idea behind `allowPartial` is that key path literal, although it might be in `PartialKeyPath` context, should still be resolved to a typed `KeyPath`, `KeyPath<SR12390, Int>` in our example, which then gets type erased by conversion to `PartialKeyPath<SR12390>`. 

One way to fix this situation might be to never "bind" key path type directly but instead have key path follow the same pattern as closures use right now (use of `resolveClosure`) and remove `KeyPathConstraint`.

Reply:
> If you look at `let _: PartialKeyPath<SR12390> = \. value` - it works just fine today because there is this bit of code which makes it possible - https://github.com/apple/swift/blob/master/lib/Sema/CSSimplify.cpp#L7534. `tryMatchRootAndValueFromType` has `allowPartial` flag which is only `true` when type is contextual, for everything else it's false - that's why when parameter type is a partial key path `simplifyKeyPathConstraint` fails.
> 


It is actually because of that part this was implemented the way it is, but I think I've missed something along the way then. Since for arg conv we don't have a contextual type, the idea was just to try to use the information from the argument type and use like a context to maybe give more information to the solver(in form of a constraint) so it could find the bindings to satisfy the key path constraint that it couldn't before. In short, use the "context" of the param to try give solver more information. Just explaining how I got to this :) 

> One way to fix this situation might be to never "bind" key path type directly but instead have key path follow the same pattern as closures use right now (use of `resolveClosure`) and remove `KeyPathConstraint`.

Ohh right, let's try that \o/ 
I'll definitely have some questions, but let's post them as it goes

And thanks for the detailed explanation @xedin :)

Reply:
> Since for arg conv we don't have a contextual type, the idea was just to try to use the information from the argument type and use like a context to maybe give more information to the solver.

In a general sense it's okay to infer argument type directly from parameter but in this particular case it's not viable to bind it because of key path semantics. Think of it this way - any information _inferred_ from constraints and other type variables is contextual by nature, non-contextual is one when type is written in the source code in a particular place. In our example argument gets a type from a parameter which is associated with one of overloads attempted by the solver, so a given argument type is viable only in context of selected overload and its parameter.

Reply:
Humm, so if I understand that right, that's probably why for the `let _: PartialKeyPath<SR12390> = \. value` the solver is capable of resolve overloads like `(overload set choice binding $T1 := Int)` for the key value $T1, where for in the case of argument this would not be possible because the context is different for each overload decl choice for the function call?

Reply:
For this expression `simplifyKeyPathConstraint` uses a different path which takes `getContextualType` into account and that works because key path type itself is not resolved.

Reply:
Hey @xedin :)
I'm getting back to this as quite understanding better somethings, but still struggling with others. 
So here I had some questions
* Not related, but what exactly the process of `salvage the system` does? To me, it was just trying to emit diagnostics but it seems like it re-solve the whole system again.
* There is a way to get the type variable associated with a root type of a keyPath expression with unresolved root e.g. `\.value` in the constraint system? I would try to get that by component ` cs.getType(keyPathExpr, 0)`, but the only component in those expr is the unresolved property ...
```
(keypath_expr type='<null>'
  (components    
    (unresolved_property decl_name='value' type='<null>'))
  (parsed_path
    (unresolved_dot_expr type='<null>' field 'value' function_ref=unapplied
      (key_path_dot_expr implicit type='<null>'))))
```
Sorry to bother with questions that way, still have some others but TBH I'm a bit confused ... trying to understand :)

Reply:
No worries!

> Not related, but what exactly the process of salvage the system does? To me, it was just trying to emit diagnostics but it seems like it re-solve the whole system again.

First round of solving is allowed to go as fast as it can (performing perf optimizations and hacks) and fail-fast in order for well-typed expressions to produce a solution quickly. If there was an error we'd go into a slower "diagnostic mode" implemented by `salvage`, the only difference is that in such mode `AllowFixes` is passed as a flag and solver is allowed to record fixes and diagnose found solutions. 

> There is a way to get the type variable associated with a root type of a keyPath expression with unresolved root e.g. \.value in the constraint system? I would try to get that by component cs.getType(keyPathExpr, 0), but the only component in those expr is the unresolved property ...

Unfortunately there is no way to get that type because it's not represented in the AST. You can refer to `ConstraintGenerator::visitKeyPathExpr` where you can see that `root` is a type variable.

Reply:
> First round of solving is allowed to go as fast as it can (performing perf optimizations and hacks) and fail-fast in order for well-typed expressions to produce a solution quickly. If there was an error we'd go into a slower "diagnostic mode" implemented by `salvage`, the only difference is that in such mode `AllowFixes` is passed as a flag and solver is allowed to record fixes and diagnose found solutions.
> 

Humm got it, thanks, this was always that way or is implemented to support the new framework? Because my impression is that it is similar in a way to `CSDiag`.

> Unfortunately there is no way to get that type because it's not represented in the AST. You can refer to ConstraintGenerator::visitKeyPathExpr where you can see that root is a type variable.
>

That's unfortunate, I'll see what I can do ... but thinking of stopping this for now, as I feel your suggestion of never bind directly and use the same pattern as closures will require quite a bit of refactoring which I'm not sure I'm able to do right now.

But thank you for all the help as always @xedin :)

Reply:
No worries! I think it might not be that bad actually because it would effectively mimic what `simplifyKeyPathConstraint` does today with an exception that it would only do it when contextual type becomes available.

Reply:
Humm, but that would mean removing the key path constraint, which maybe can have other implications? Anyways I'll take another look soon :) Thanks @xedin

Reply:
> I think it might not be that bad actually because it would effectively mimic what simplifyKeyPathConstraint does today with an exception that it would only do it when contextual type becomes available.

@xedin  
I've been trying to implement something like this but TBH somethings got me confused:
* `simplifyKeyPathConstraint` rely on information about `root` and `value` types as well as component information in `componentTypeVars` that are computed in `CSGen` and passed along in the keyPath constraint to be able to resolve the key path type. So in order to do logic to resolve the key path type on a separated function like that `resolveClosure` (in that case let's say `resolveKeyPath`), we should move `CSGen` logic(that create type variables for key path components like the resolve closure creates them for the function body) to that `resolveKeyPath` function? Or get that information from the key path constraint in some way? Or am I getting this all wrong 🤣 ? 
* Move that logic from that will make `simplifyKeyPathConstraint` not necessary anymore? So, the key path constraint will not be needed?
* I got confused with all that and went on looking for another way which is instead of doing like this code here trying to get direct info from the param, just try to get the root and let the `simplifyKeyPathConstraint` do the rest and resolve the type, which seems to works fine ... do you think that would make sense? I'll send a commit with that for you to see.

Reply:
This 03af8ec66d192e249efa7cc2183e6069bed54750 or some way to provide this information to the solver only when attempting an overload choice involving a partial key path?

Reply:
I think it should follow `resolveClosure` playbook and generate a placeholder type variable for a key path, once that type variable gets a binding from contextual, `resolveKeyPath` should perform constraint generation (which used to be done in CSGen before). The reason why key path constraint exists is related to the fact that we need to wait until all of the components are resolved to figure out a _capability_ of key path (read-only, writable, reference writable). I think that would no longer be necessary because we'd know what capability is expected by context and we just need to validate that after each component is resolved.

What I had in mind for components is to connect them into a directed chain via `one-way` constraints (result of the previous component is collected to the next via such constraint) that would make sure that all of the components are resolved in order. We could add a new `key path component` constraint which acts just like `one way` constraint with some special logic that makes sure that capability of context is maintained.

This would also solve a problem where `key path` constraint is re-activated N times before all of the components are resolved, which creates wasted work.

Reply:
Awesome @xedin! I think I'm finally understanding it.
Thanks again for such a detailed explanation :)

Reply:
Hey @xedin, 
After some time working on this with not much progress, I think I'm going to move it down in the queue and work on some other issues probably...
Sorry for the letdown, but thank you so much for all the help and detailed explanations! :)

Reply:
No worries! I think it would be useful to write it all down in SR for posterity.

Reply:
> No worries! I think it would be useful to write it all down in SR for posterity.

Sure, On it! :)

Reply:
Done! Thanks for all @xedin :)